design/dsl/request_test.go

```diff
@@ -0,0 +1,131 @@
+package dsl_test
+
+import (
+	"testing"
+
+	"reflect"
+
+	"goa.design/goa.v2/design"
+	. "goa.design/goa.v2/design/dsl"
+	"goa.design/goa.v2/eval"
+)
+
+func TestRequest(t *testing.T) {
+	var RequestType = Type("Request", func() {
+		Description("Optional description")
+		Attribute("required", design.String)
+		Attribute("name", design.String)
+		Required("required")
+	})
+	cases := map[string]struct {
+		DSL         func()
+		Assert      func(t *testing.T, req design.DataType)
+		ExpectError bool
+	}{
+		"DefinedUserType": {
+			DSL: func() {
+				Request(RequestType, func() {
```

Comment:
What this does is build a user type of type object. So you can get to the attributes with something like (assuming everything is correct, you probably need to check for nil and correct types in real code):
```go
obj := req.(*design.UserTypeExpr).Type.(design.Object)
for name, att := range obj {
    // name contains the attribute name
    // att contains the attribute
}
```

Hmm I see you do this below so I'm probably not understanding the question...